quote.py

```diff
@@ -3,48 +3,102 @@
 import random
 from halibot import HalModule
 
+sr = random.SystemRandom()
+
 class Quote(HalModule):
 
-	quotes = []
+	# Configuration defaults
+	remove_votes = 3
+	max_history = 10
 
 	def init(self):
-		self.path = self.config['quotes-path']
+		self.quotes = []
+		self.path = self.config.get('quotes-path')
+		self.remove_votes = self.config.get('remote-votes', self.remove_votes)
+		self.max_history = self.config.get('max-history', self.max_history)
 		print('using quotes path:', self.path)
 
-		with open(self.path, 'r') as f:
-			self.quotes = [x.rstrip() for x in f.readlines()]
+		if self.path is not None:
+			try:
+				with open(self.path, 'r') as f:
+					self.quotes.extend([x.rstrip() for x in f if x.rstrip()])
+			except IOError, OSError:
+				pass
+
+		self.rand_quotes = {}  # pattern -> shuffled lists
+		self.rem_vote_map = {}  # quote -> set(usernames)
 
 	def quote(self, pattern=""):
-		if len(pattern) > 0:
-			try:
-				expr = re.compile(pattern)
-			except re.error as e:
-				return 'Invalid pattern: ' + str(e)
-				
-			ls = [q for q in self.quotes if re.search(expr, q)]
+		if pattern in self.rand_quotes:
+			ls = self.rand_quotes[pattern]
+			quote = ls.pop()
+			if not ls:
+				del self.rand_quotes[pattern]
 		else:
-			ls = self.quotes
+			if len(pattern) > 0:
+				try:
+					expr = re.compile(pattern)
+				except re.error as e:
+					return 'Invalid pattern: %s' % (e,)
 
-		if len(ls) > 0:
-			return random.choice(ls)
-		else:
-			return 'No quotes found with that pattern :('
+				ls = sr.shuffle([q for q in self.quotes if expr.search(q)])
+			else:
+				ls = sr.shuffle(self.quotes)
+
+			if not ls:
+				return 'No quotes found with that pattern :('
+
+			if len(self.rand_quotes) >= self.max_history:
+				for i in range(len(self.rand_quotes) - self.max_history + 1):
+					self.rand_quotes.remove(sr.choice(self.rand_quotes))  # XXX LRU?
+
+			self.rand_quotes[pattern] = ls
+			quote = self.quote(pattern)  # Recursion!
+
+		return quote
 
 	def quoteadd(self, quote):
 		self.quotes.append(quote)
 
-		with open(self.path, 'a') as f:
-			print(quote, file=f)
+		if self.path is not None:
+			try:
+				with open(self.path, 'a') as f:
+					print(quote, file=f)
+			except IOError, OSError as e:
+				return 'Couldn\'t add quote! I\'ll remember it for now, though :) (%s)' % (e,)
 
 		return 'Added :)'
 
+	def quotedel(self, quote, nick):
+		if quote not in self.quotes:
```

Comment:
Thought on this a little longer, might be useful to print a useful message as confirmation, either that it matched too many quotes, or a truncated version of the quote to confirm.